api/include/opentelemetry/trace/propagation/http_trace_context.h

```diff
@@ -0,0 +1,313 @@
+static const int kSpanIdBytes                = 16;
+static const int kTraceFlagBytes             = 2;
+static const int kTraceDelimiterBytes        = 3;
+static const int kHeaderSize =
+    kVersionBytes + kTraceIdBytes + kSpanIdBytes + kTraceFlagBytes + kTraceDelimiterBytes;
+static const int kTraceStateMaxMembers    = 32;
+static const int kHeaderElementLengths[4] = {2, 32, 16, 2};
+
+// The HttpTraceContext provides methods to extract and inject
+// context into headers of HTTP requests with traces.
+// Example:
+//    HttpTraceContext.inject(setter,&carrier,&context);
+//    HttpTraceContext.extract(getter,&carrier,&context);
+template <typename T>
+class HttpTraceContext : public HTTPTextFormat<T>
+{
+public:
+  // Rules that manages how context will be extracted from carrier.
+  using Getter = nostd::string_view (*)(const T &carrier, nostd::string_view trace_type);
+
+  // Rules that manages how context will be injected to carrier.
+  using Setter = void (*)(T &carrier,
+                          nostd::string_view trace_type,
+                          nostd::string_view trace_description);
+
+  void Inject(Setter setter, T &carrier, const context::Context &context) noexcept override
+  {
+    SpanContext span_context = GetCurrentSpan(context)->GetContext();
+    if (!span_context.IsValid())
+    {
+      return;
+    }
+    InjectImpl(setter, carrier, span_context);
+  }
+
+  context::Context Extract(Getter getter,
+                           const T &carrier,
+                           context::Context &context) noexcept override
+  {
+    SpanContext span_context    = ExtractImpl(getter, carrier);
+    nostd::string_view span_key = "current-span";
+    nostd::shared_ptr<Span> sp{new DefaultSpan(span_context)};
+    return context.SetValue(span_key, sp);
+  }
+
+  static Span *GetCurrentSpan(const context::Context &context)
+  {
+    const nostd::string_view span_key = "current-span";
+    context::Context ctx(context);
+    nostd::shared_ptr<Span> span = nostd::get<nostd::shared_ptr<Span>>(ctx.GetValue(span_key));
+    return (span.get());
+  }
+
+  static TraceId GenerateTraceIdFromString(nostd::string_view trace_id)
+  {
+    const char *trc_id = trace_id.begin();
+    uint8_t buf[kTraceIdBytes / 2];
+    int tmp;
+    for (int i = 0; i < kTraceIdBytes; i++)
+    {
+      tmp = CharToInt(*trc_id);
+      if (tmp < 0)
+      {
+        for (int j = 0; j < kTraceIdBytes / 2; j++)
+        {
+          buf[j] = 0;
+        }
+        return TraceId(buf);
+      }
+      if (i % 2 == 0)
+      {
+        buf[i / 2] = tmp * 16;
+      }
+      else
+      {
+        buf[i / 2] += tmp;
+      }
+      trc_id++;
+    }
+    return TraceId(buf);
+  }
+
+  static SpanId GenerateSpanIdFromString(nostd::string_view span_id)
+  {
+    const char *spn_id = span_id.begin();
+    uint8_t buf[kSpanIdBytes / 2];
+    int tmp;
+    for (int i = 0; i < kSpanIdBytes; i++)
+    {
+      tmp = CharToInt(*spn_id);
```

Comment:
Does it work to use `spn_id[i]` here and remove the `spn_id++` at the end of the loop?

Having just one loop index (`i`) that we use everywhere avoids any chance that things get out of sync.

Reply:
I have replaced it